src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -754,13 +752,16 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             _incomingBuffer.Discard(frameHeader.PayloadLength);
         }
 
-        internal async Task FlushAsync(CancellationToken cancellationToken = default)
-        {
-            await StartWriteAsync(0, cancellationToken).ConfigureAwait(false);
-            FinishWrite(FlushTiming.Now);
-        }
+        internal Task FlushAsync(CancellationToken cancellationToken) =>
+            InvokeLockedAsync(0, this, (thisRef, writeBuffer) => thisRef.FinishWrite(writeBuffer.Length, FlushTiming.Now), null, cancellationToken);
 
-        private async ValueTask<Memory<byte>> StartWriteAsync(int writeBytes, CancellationToken cancellationToken = default)
+        /// <summary>Performs a write operation serialized via the <see cref="_writerLock"/>.</summary>
+        /// <param name="writeBytes">The number of bytes to be written.</param>
+        /// <param name="state">The state to pass through to the callbacks.</param>
+        /// <param name="lockedAction">The action to be invoked while the writer lock is held and that actually writes the data to the provided buffer.</param>
+        /// <param name="failedAcquireAction">The action to invoke if waiting for the write lock is canceled.</param>
+        /// <param name="cancellationToken">The cancellation token to use while waiting.</param>
+        private async Task InvokeLockedAsync<T>(int writeBytes, T state, Action<T, Memory<byte>> lockedAction, Action<T>? failedAcquireAction = null, CancellationToken cancellationToken = default)
```

Comment:
Can we call this something more descriptive, like `PerformWriteAsync` or something like that?

Reply:
Previously, callers would deal with failure cases here with just try/catch. Can we continue to do that rather than adding `failedAcquireAction`? 

Note that I believe the action itself should never throw. Perhaps we should assert that too.

Reply:
> Can we call this something more descriptive, like PerformWriteAsync or something like that?

Sure

> Can we fold the call to FinishWrite in to this routine? That would avoid having to pass the writeBytes again to FinishWrite.

I think so; we can just have the delegate return the flush mode that's desired.

> Previously, callers would deal with failure cases here with just try/catch. Can we continue to do that rather than adding failedAcquireAction?

If we're not concerned about what might go wrong if the delegate threw an exception (e.g. an OOM), then sure.  It sounds like you're saying that's the case.

Reply:
> If we're not concerned about what might go wrong if the delegate threw an exception (e.g. an OOM), then sure. It sounds like you're saying that's the case.

I don't even think they can OOM, but I haven't checked rigorously...

Reply:
> I think so; we can just have the delegate return the flush mode that's desired.

Either that or just pass the flush mode to InvokeLockedAsync. I'm not sure there's any case where we can't determine it before making this call.